### ISSUE TITLE
DeepSeek-V4 HC mixer params downcast to bf16

### ISSUE THREAD

**Issue by jQizhang** [NVIDIA-NeMo/Automodel#2143]
**Describe the bug**

For models that declare module names in `_keep_in_fp32_modules_strict` (e.g. DeepSeek-V4's head-channel mixers — `attn_hc`, `ffn_hc`, `hc_head`), Automodel's `cast_model_to_dtype` silently downcasts those parameters to bf16 when the model is sharded with FSDP2. The framework emits a single `WARNING` line and proceeds:

```
Model parameters are DTensors (FSDP2) — skipping fp32 parameter restoration
for keywords=[...]. Only buffers will be restored to fp32. FSDP2 requires
uniform dtype within each parameter group.
```

**Steps/Code to reproduce bug**

1. Use any DSV4 fine-tune YAML, e.g. `examples/llm_finetune/deepseek_v4/deepseek_v4_flash_hellaswag.yaml`.
2. After `recipe.setup()`, dump the dtype of HC parameters:

   ```python
   from collections import Counter

   c = Counter()
   for n, p in model.named_parameters():
       if any(k in n for k in ("attn_hc", "ffn_hc", "hc_head")):
           c[(n, str(p.dtype))] += 1
   for k, v in c.items():
       print(k, v)
   ```

Observed: every HC parameter is `torch.bfloat16`, accompanied by the `"skipping fp32 parameter restoration"` warning. The `_keep_in_fp32_modules_strict` declaration on the model class has no runtime effect.

**Expected behavior**

Parameters whose module names are listed in `_keep_in_fp32_modules_strict` should remain in `torch.float32` at runtime, regardless of whether the model is wrapped with FSDP2 or not. For DSV4 this means all `attn_hc.{fn,base,scale}`, `ffn_hc.{fn,base,scale}`, and `hc_head.{hc_fn,hc_base,hc_scale}` tensors should be `torch.float32` after `recipe.setup()`, matching both the safetensors checkpoint dtype and vLLM's hard requirement. The same expectation applies to buffer names (e.g. `e_score_correction_bias`) declared in that list.

**Additional context**

We first surfaced this bug through an unusually large training-vs-inference mismatch KL in NeMo-RL GRPO runs on DSV4. Because vLLM runs the HC mixers in fp32 while Automodel-trained policy was silently running them in bf16, the same prompts produced noticeably different logprob distributions across the two runtimes:

| Configuration | Train/inference mismatch KL |
|---|---:|
| Before fix (Automodel HC = bf16, vLLM HC = fp32) | ~0.15 – 0.20 |
| After fix (Automodel HC = fp32, vLLM HC = fp32)  | ~0.03 |

**Comment by HuiyingLi** [comment]
Hi @jQizhang — thanks for the detailed writeup. We're aware of this and have our own DSV4 parity work in progress.

In addition to the HC mixers (attn_hc, ffn_hc, hc_head) you flagged, we also found the following components benefit from staying / running in FP32 to match the released checkpoint and the dsv4flash reference behavior:
  - MoE gate scoring (Gate.forward for the sqrtsoftplus path) — set via BackendConfig.gate_precision=torch.float32.
  - DeepseekV4RotaryEmbedding cos/sin output — reference keeps angles fp32 throughout the complex-mul rope;
  - _apply_partial_rope_interleaved rotation multiply — bf16 pair-multiply on rope adds ~10⁻³ rel-error per element; fp32 internal compute (downcast at end) eliminates it.
  - eager_attention_with_sink softmax — reference TileLang sparse_attn keeps softmax fp32; bf16 softmax over 2048 keys is ~3× further from FP64 ground truth than ref's path.
  - HC residual update (post * attn_out + comb @ x) inside DeepseekV4Block.forward 
  - lm_head projection 

Of these, the eager-attention-with-sink → reference sparse_attn precision gap is the largest single contributor we've measured so far. The HC mixer one is the only one we've connected to a downstream training metric so far thank you! The others are precision-correct against the reference but we haven't measured a comparable train→inference impact for them — would be very curious whether your GRPO setup sees movement on any of them too?

Do you happen to have a more thorough study of dtype sensitivity across DSV4 we could reference? It'd help us prioritize which of the above are worth landing as core fixes vs leaving as opt-in precision tweaks.

PRs very welcome! Let us know if you'd like to take it. Thank you so much! Appreciate it!

**Comment by jQizhang** [comment]
Hi @HuiyingLi  thanks a lot for the detailed suggestions.

I realized my original issue description was a bit inaccurate: The GRPO results I shared actually incorporated two distinct fixes: the MHC FP32 casting and a fix for a HC comb transpose issue (which I have documented in a separate issue: #2154 )

To better understand the individual impact of these factors, I ran a small offline ablation. So far, it looks like the HC comb transpose issue is the main contributor to the larger mismatch KL, while the MHC FP32 change alone has only a very small effect in this setup.

Test Case: 165 input tokens / 2048 output tokens.

| Variant | KL (Native BF16 || Automodel) |
| Baseline | 0.265243 |
| MHC dtype in FP32 | 0.264759 |
| HC comb transpose | 0.038195 |
| MHC fp32 + transpose | 0.038090

So based on this ablation, I would not claim that the MHC FP32 path is a major contributor to the KL gap in my setup. However, since these modules are already explicitly listed in [_keep_in_fp32_modules_strict](https://github.com/NVIDIA-NeMo/Automodel/blob/1581fd0a101bd2f616f23b737ec8c6a4c02c48c0/nemo_automodel/components/models/deepseek_v4/model.py#L479 ), maybe we should also enforce FP32 here?

I’d appreciate your suggestions on this! Thanks!

**Comment by HuiyingLi** [comment]
@jQizhang , thank you so much for sharing. Yeah I've also tried the mHC 32 cast and didn't see much improvement so I didn't hurry to add it in. But I think they should be added. If you have time to help with the upcast that'd be really helpful.
The transpose issue is a real bug though. I will reply in the other issue separately. 
Again thank you so much for your effort!

**Comment by yiakwy-xpu-ml-framework-team** [comment]
@jQizhang embeded and tid2eid also wrongly represented in bf16. （found when cast bf16 to fp8/fp32/bf16 mix precision model）

```
"embed", "tid2eid" ：int64
"hc_attn_fn"，"hc_ffn_fn"，“compressor.ape”, "hc_head_fn"： fp32
```

We found "embed", "tid2eid" are represented with bf16.

**Comment by khazic** [comment]
Opened PR #2216 to address the dtype alignment discussed here: https://github.com/NVIDIA-NeMo/Automodel/pull/2216

The PR covers the reference-sensitive DeepSeek-V4 dtype paths from the official DeepSeek-V4-Flash inference implementation:

- Keeps HC mixer/head parameters in fp32 (attn_hc.*, ffn_hc.*, hc_head.*).
- Keeps lm_head, attention sink, and compressor/indexer ape/wkv/wgate parameters in fp32.
- Runs RoPE pair multiply, lm_head, compressor, and indexer paths with fp32 internal compute where the reference does so.
- Changes hash gate tid2eid to int32 and preserves it across dtype casts.
- Adds dtype-aware FSDP2 sharding for DeepSeek-V4 so fp32 and bf16 parameters are not wrapped in the same shard unit.

One note on embed: I did not change embed.weight to int64/fp32 because the official ParallelEmbedding.weight is a floating parameter created under the model default dtype, not an integer lookup table. The integer table called out in the thread is covered by tid2eid.

**Comment by yiakwy-xpu-ml-framework-team** [comment]
@khazic Hi I want to udpated a script to convert bf16 from Automodel to fp8 (correctly running in SGLang). Where should I put ?

**Comment by HuiyingLi** [comment]
@yiakwy-xpu-ml-framework-team Thank you so much! I'd suggest tools/
